packages/backend-function/src/function_env_type_generator.ts

```diff
@@ -2,58 +2,61 @@ import fs from 'fs';
 import { staticEnvironmentVariables } from './static_env_types.js';
 import path from 'path';
 import os from 'os';
-import { FunctionProps } from './factory.js';
 
 /**
- * Generates a type definition file for environment variables
+ * Generates a typed process.env shim for environment variables
  */
 export class FunctionEnvironmentTypeGenerator {
   private typeDefFilePath: string;
-  private dynamicEnvironmentVariables: string[] = [];
 
   /**
-   * Initialize type definition file name and location
+   * Initialize typed process.env shim file name and location
    */
   constructor(
     functionName: string,
-    functionEnvironmentVariables: FunctionProps['environment']
+    private functionEnvironmentVariables: string[]
   ) {
     this.typeDefFilePath = `${process.cwd()}/.amplify/function-env/${functionName}.ts`;
-
-    for (const envName in functionEnvironmentVariables) {
-      this.dynamicEnvironmentVariables.push(envName);
-    }
   }
 
   /**
-   * Generate a type definition file
+   * Generate a typed process.env shim
    */
-  generateTypeDefFile() {
+  generateTypedProcessEnvShim() {
+    const lambdaEnvVarTypeName = 'LambdaProvidedEnvVars';
+    const definedEnvVarTypeName = 'DefinedEnvVars';
     const declarations = [];
     const typeDefFileDirname = path.dirname(this.typeDefFilePath);
 
     if (!fs.existsSync(typeDefFileDirname)) {
       fs.mkdirSync(typeDefFileDirname, { recursive: true });
     }
 
+    // Add Lambda runtime environment variables
+    declarations.push(`type ${lambdaEnvVarTypeName} = {${os.EOL}`);
     for (const key in staticEnvironmentVariables) {
       const comment = `/** ${staticEnvironmentVariables[key]} */`;
       const declaration = `${key}: string;`;
 
-      declarations.push(comment + os.EOL + declaration);
+      declarations.push(comment + os.EOL + declaration + os.EOL + os.EOL);
     }
+    declarations.push(`};${os.EOL + os.EOL}`);
 
-    this.dynamicEnvironmentVariables.forEach((envName) => {
+    // Add defined environment variables
+    declarations.push(`type ${definedEnvVarTypeName} = {${os.EOL}`);
+    this.functionEnvironmentVariables.forEach((envName) => {
       const declaration = `${envName}: string;`;
 
-      declarations.push(declaration);
+      declarations.push(declaration + os.EOL);
     });
+    declarations.push(`};${os.EOL + os.EOL}`);
 
     const content =
       `/** Lambda runtime environment variables, see https://docs.aws.amazon.com/lambda/latest/dg/configuration-envvars.html#configuration-envvars-runtime */${os.EOL}` +
-      `export const env = process.env as {${os.EOL}` +
-      declarations.join(os.EOL + os.EOL) +
-      `${os.EOL}};`;
+      `export const env = process.env as ${
+        lambdaEnvVarTypeName + ' & ' + definedEnvVarTypeName
+      };${os.EOL + os.EOL}` +
+      declarations.join('');
```

Comment:
```suggestion
      `export const env = process.env as ${lambdaEnvVarTypeName} & ${definedEnvVarTypeName};${EOL}${EOL}${declarations.join('')}`
```
Also applies elsewhere in this file; use only template strings instead of combining string concatenation and template strings

Reply:
Also, could we clean up some of the `os.EOL` stuff in other places if we did `declarations.join(EOL)` here?

Reply:
Ah you're right, I took it off and started manually adding `EOL`, should've seen we can do `declarations.join(EOL` at least afterwards.